### ISSUE TITLE
SeqConcat does not support indexer but no way to determine if a Seq is internally using SeqConcat.

### ISSUE THREAD

**Issue by glen-nicol** [louthy/language-ext#1217]
Title mostly says it all. I have a utility function that does a binary search on a Seq but if the caller passes a Seq that involves a Concat then I get NotSupportedException when indexing into the middle. I don't see a way to determine if indexing is supported or not from a Seq instance (The internal seq is marked internal and thus I can't inspect it). 

Currently doing this which is unfortunate since it involves catching an exception:

```csharp
Prelude.Try(() => seq[middle])
.ToEither()
.IfLeft(e => e is NotSupportedException ? seq.Skip(middle).Head : throw e)
```

improvement ideas (not necessarily mutually exclusive):
1. Add property to indicate fast indexing is supported
2. SeqConcat (and maybe other implementations that throw) can use the Skip().Head method internally instead of throwing
3. If 2 isn't tenable then option 1 would at least allow an extension method to be made that uses the indexer and then fallsback on skip method.

Thoughts on those or existing alternatives?

**Comment by louthy** [comment]
I’ll take a look tomorrow, as a workaround you can force strictness through .Strict() - if you’re indexing then I guess you don’t need laziness?

**Comment by glen-nicol** [comment]
Thoughts on this implementation for the SeqConcat indexer?

```csharp
public A this[int index]
{
    get
    {
        if(index < 0)
        {
            throw new IndexOutOfRangeException("Negative index is invalid.");
        }

        var runningIndex = index;

        foreach (var s in ms)
        {
            if (s.Count < runningIndex)
            {
                return s[runningIndex];
            }

            runningIndex -= s.Count;
        }

        throw new IndexOutOfRangeException("Index was outside the bounds of the Sequence");
    }
}
```

I am working on tests but I am nearly certain that is correct.

**Comment by louthy** [comment]
This is now fixed and released as version `4.4.3`

There's now also a 'safe indexer' method called `At` on [`Seq`](https://louthy.github.io/language-ext/LanguageExt.Core/Immutable%20Collections/Seq/index.html#Seq_1_At_0) and [`Lst`](https://louthy.github.io/language-ext/LanguageExt.Core/Immutable%20Collections/List/index.html#Lst_1_At_0) which will return `Option<A>`.

> Thoughts on this implementation for the SeqConcat indexer?

It would work, but it forces all of the (current) sequence to stream-in (it calls `Count` before indexing), which for lazy sequences would read more than is needed to perform the index operation.

So if we have a  concatted sequence, the first part of which is lazy:
```csharp
var s = Seq(Range(1, 1000)).Concat(...);
``` 
And, we call `s[0]` to get the first item, then we'd end up streaming 1000 items just to get the first one.

I've updated `ISeqInternal` to have a 'safe' indexer, called `At` that returns `Option<A>`.  If it can index, it will, otherwise it returns `None`.  This allows for a more controlled index and handling of out-of-range errors.   

The new `SeqConcat` loop looks a bit like yours:

```c#
        public Option<A> At(int index)
        {
            if (index < 0) return default;
            var ms1 = ms;
            while (!ms1.IsEmpty)
            {
                var head = ms1.Head;
                var r = head.At(index);
                if (r.IsSome) return r;
                index -= head.Count;
                ms1 = ms1.Tail;
            }
            return default;
        }
```
The difference is we only reduce the index by `Count` _after_ we know we've just scanned over it (and therefore all items within have definitely been streamed in).  This stops any overeagerness.

You could have done similar by attempting to index (without the `Count` check), catching the `IndexOutOfRangeException` and only then reducing the index by `Count`.

**Comment by glen-nicol** [comment]
Yes very true. I guess I was thinking if someone is using the indexer then Seq.Count was probably already called and streamed the entire Seq into memory. But you're right that may not always be the case and your implementation should work for both cases. 

Thanks so much for the quick response!
